srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,1405 @@
+			);
+
+			if !Self::is_voter(&who) {
+				// first time voter. Reserve bond.
+				T::Currency::reserve(&who, T::VotingBond::get())
+					.map_err(|_| "voter can not pay voting bond")?;
+			}
+			// Amount to be locked up.
+			let locked_balance = value.min(T::Currency::total_balance(&who));
+
+			// lock
+			T::Currency::set_lock(
+				MODULE_ID,
+				&who,
+				locked_balance,
+				T::BlockNumber::max_value(),
+				WithdrawReasons::all(),
+			);
+			<StakeOf<T>>::insert(&who, locked_balance);
+			<VotesOf<T>>::insert(&who, votes);
+		}
+
+		/// Remove `origin` as a voter. This removes the lock and returns the bond.
+		#[weight = SimpleDispatchInfo::FixedNormal(500_000)]
+		fn remove_voter(origin) {
+			let who = ensure_signed(origin)?;
+
+			ensure!(Self::is_voter(&who), "must be a voter");
+
+			Self::do_remove_voter(&who, true);
+		}
+
+		/// Report `target` for being an defunct voter. In case of a valid report, the reporter is
+		/// rewarded by the bond amount of `target`. Otherwise, the reporter itself is removed and
+		/// their bond is slashed.
+		///
+		/// A defunct voter is defined to be:
+		///   - a voter who's current submitted votes are all invalid. i.e. all of them are no
+		///     longer a candidate nor an active member.
+		///
+		/// # <weight>
+		/// O(NLogM) given M current candidates and N votes for `target`.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(500_000)]
+		fn report_defunct_voter(origin, target: <T::Lookup as StaticLookup>::Source) {
+			let reporter = ensure_signed(origin)?;
+			let target = T::Lookup::lookup(target)?;
+
+			ensure!(reporter != target, "cannot report self");
+			ensure!(Self::is_voter(&reporter), "reporter must be a voter");
+
+			// Checking if someone is a candidate and a member here is O(LogN), making the whole
+			// function O(MLonN) with N candidates in total and M of them being voted by `target`.
+			// We could easily add another mapping to be able to check if someone is a candidate in
+			// `O(1)` but that would make the process of removing candidates at the end of each
+			// round slightly harder.
+			//
+			// Note that for now we have a bound of number of votes (`N`).
+			let valid = Self::is_defunct_voter(&target);
+			if valid {
+				// reporter will get the voting bond of the target
+				T::Currency::repatriate_reserved(&target, &reporter, T::VotingBond::get())?;
+				// remove the target. They are defunct.
+				Self::do_remove_voter(&target, false);
+			} else {
+				// slash the bond of the reporter.
+				let imbalance = T::Currency::slash_reserved(&reporter, T::VotingBond::get()).0;
+				T::BadReport::on_unbalanced(imbalance);
+				// remove the reporter.
+				Self::do_remove_voter(&reporter, false);
+			}
+			Self::deposit_event(RawEvent::VoterReported(target, reporter, valid));
+		}
+
+
+		/// Submit oneself for candidacy.
+		///
+		/// A candidate will either:
+		///   - Lose at the end of the term and get slashed.
```

Comment:
```suggestion
		///   - Lose at the end of the term and forfeit their deposit.
```

kicking everyone except the winners is no good. we really want to retain a number of well-backed runners-up while kicking the others. two options spring to mind:

- re-run phragmen with all stakes on the losers to determine N runners-up.
- re-run phragmen with all stakes to determine seats + N winners + runners-up, then remove the winners from the set to give runners-up.

Reply:
Option B is what i was doing before (in 1 run, not twice) and we sorta decided to remove it as it not really a runner up set. 

Nonetheless it might be the only way forward. I also think it could, in practice, always yield the same results as option A. Either way I will have a chat with Al and the research team before taking this forward.

Reply:
yeah but before was no good as we corrupted the winning set in order to find runners-up. option B is different in that we run phragmen twice with a different size of winning set, thereby avoiding such conflicts.

Reply:
brought back the runner ups.

Reply:
(runners up) :-)